get_operation() right-trim underflows size_t and writes below its buffer

Found by a `-Wextra` sweep over every TU while working on #267. cc370 is
GCC 3.4.6, which has no `-Wtype-limits` (4.3+) — the tautological-comparison
warning lives under `-Wextra` there. This is the only hit outside the two
`dsapi.c` handlers #267 names:

```
src/jobsapi.c:1845: warning: comparison of unsigned expression >= 0 is always true
```

Unlike #267's dead checks, this one is not inert. It writes.

## The defect

`get_operation()` (`src/jobsapi.c:1805`) copies the operation field of a JCL
card and then right-trims it:

```c
size_t i = 0;
while (i < 4 && i < op_size - 1 && pos < end && *pos && *pos != ' ') {
    op[i++] = *pos++;
}
op[i] = '\0';

// Right trim
for (i--; i >= 0; i--) {
    if (isspace((unsigned char)op[i])) {
        op[i] = '\0';
    } else {
        break;
    }
}
```

`i` is `size_t`. When the operation field comes out **empty**, `i` is 0 and
`i--` wraps to `SIZE_MAX` — the `i >= 0` guard cannot stop it, which is what
the warning says. `op[SIZE_MAX]` is `op[-1]`: the byte immediately below a
5-byte array on the caller's stack frame.

It is not a stray read. If that byte is a blank — 0x40, and blank-padded JCL
buffers sit nearby — it is overwritten with `'\0'` and the loop steps down to
`op[-2]`, `op[-3]`, … zeroing every blank it finds until it meets a non-blank
byte. A run of blanks below the array is a run of zeroed stack.

## What reaches it

`find_job_card_range()` calls `get_operation()` for **every** line of
submitted JCL that is at least 3 characters long and starts with `//`.
The operation field is empty for ordinary cards:

- `// ` — `//` followed by blanks
- `//NAME` followed by blanks and nothing else

A bare `//` — the delimiter card that ends most jobs — is *not* affected:
`line_len < 3` returns before the trim. Two characters of luck.

## Fix

Walk the length, not the index — the loop cannot underflow if it never
decrements past zero:

```c
while (i > 0 && isspace((unsigned char)op[i - 1])) {
    op[--i] = '\0';
}
```

The trim is worth keeping rather than deleting: the copy loop stops at `' '`,
so a blank can never land in `op`, but tab/CR/LF can, and `isspace()` matches
those.

## Note

The whole-tree `-Wextra` sweep is cheap and finds this class in one pass —
worth repeating occasionally. Filter hard: `-Wextra` reports 61 other
warnings on this codebase, none of them this class.

--- a/src/dsapi.c
+++ b/src/dsapi.c
@@ -1248,7 +1248,12 @@ dslist_in_page(const DSLIST *ds, const char *start_key, int have_start,
 
 int datasetListHandler(Session *session)
 {
-	unsigned	rc		= 0;
+	/* signed, and that is the whole point: `if ((rc = http_printf(...)) < 0)`
+	   is dead code on an unsigned rc, so every write below went unchecked and a
+	   client that disconnected mid-listing had the rest of the catalog formatted
+	   and sent after it (#267).  Only rc changes -- the counters below are
+	   compared and bounded as unsigned on purpose. */
+	int		rc		= 0;
 	unsigned	count		= 0;
 	unsigned	first		= 1;
 	unsigned	i		= 0;
@@ -1297,7 +1302,7 @@ int datasetListHandler(Session *session)
 			/* goto quit like every other exit here: dslist is still NULL at
 			   this point, so a bare return would be correct today and wrong
 			   the moment anything is allocated above it. */
-			rc = (unsigned) handle_error(session, ERR_INVALID_PARAM,
+			rc = handle_error(session, ERR_INVALID_PARAM,
 					"Dataset level is too long");
 			goto quit;
 		}
@@ -2291,7 +2296,11 @@ member_scan(Session *session, FILE *fp, const char *start_key, int start_after,
 
 int memberListHandler(Session *session)
 {
-	unsigned	rc		= 0;
+	/* signed for the reason spelled out in datasetListHandler() (#267).  Here it
+	   is the header and trailer writes that were unchecked; the directory walk
+	   itself is already guarded, because member_scan() returns int and the
+	   caller below tests it. */
+	int		rc		= 0;
 	unsigned	emitted		= 0;
 	unsigned	maxitems	= 0;
 	int		truncated	= 0;
@@ -2337,7 +2346,7 @@ int memberListHandler(Session *session)
 	   why the wrong one would survive review and then break when something is
 	   allocated above it. */
 	if (!normalize_dsn(dsname, dsn_buf, sizeof(dsn_buf))) {
-		rc = (unsigned) handle_error(session, ERR_INVALID_PARAM,
+		rc = handle_error(session, ERR_INVALID_PARAM,
 				"Dataset name is too long");
 		goto quit;
 	}
@@ -2440,7 +2449,7 @@ int memberListHandler(Session *session)
 	scanned = member_scan(session, fp, skipping ? start_key : NULL,
 			start_after, pattern_key, have_pattern, maxitems);
 	if (scanned < 0) {
-		rc = (unsigned) scanned;
+		rc = scanned;
 		goto quit;
 	}
 

--- a/src/jobsapi.c
+++ b/src/jobsapi.c
@@ -1841,13 +1841,18 @@ get_operation(const char *line, char *op, size_t op_size)
     }
     op[i] = '\0';
 
-    // Right trim
-    for (i--; i >= 0; i--) {
-        if (isspace((unsigned char)op[i])) {
-            op[i] = '\0';
-        } else {
-            break;
-        }
+    /* Right trim.  `for (i--; i >= 0; i--)` was an underflow, not a loop
+       (#353): i is size_t, so on an empty operation field i-- is SIZE_MAX and
+       op[SIZE_MAX] is op[-1] -- the byte below the array.  A blank there was
+       overwritten with '\0' and the walk carried on down the caller's frame
+       for as long as it kept finding blanks.  `// ` and a name-only card both
+       produce an empty field, and find_job_card_range() calls this for every
+       line of submitted JCL.  Walking the length cannot step past zero.
+
+       The trim itself stays: the copy loop above stops at ' ', so a blank can
+       never land in op, but tab/CR/LF can and isspace() matches those. */
+    while (i > 0 && isspace((unsigned char)op[i - 1])) {
+        op[--i] = '\0';
     }
 }
 

--- a/tests/curl-jobs.sh
+++ b/tests/curl-jobs.sh
@@ -385,6 +385,47 @@ test_submit_notify_sysuid_trailing_param() {
 	fi
 }
 
+test_submit_padded_null_statement() {
+	echo ""
+	echo "--- Submit Job: blank-padded null statement in the deck (issue #353) ---"
+
+	# get_operation() right-trimmed with `for (i--; i >= 0; i--)` on a size_t, so
+	# an empty operation field wrapped i to SIZE_MAX and the trim wrote below its
+	# own five-byte buffer, zeroing every blank it met in the caller's frame. A
+	# "//" card padded to column 80 is exactly that case -- and it is what a fixed
+	# RECFM=F deck looks like, so it ran on ordinary submits, this suite's own #130
+	# test included. The corruption is silent, so this is a regression guard rather
+	# than a reproduction: it asserts the deck still parses and the job card is
+	# still found, which a lost card would answer 400.
+	local jcl
+	jcl=$(printf '%-80s\n%-80s\n%-80s\n' \
+		'//NULLST   JOB CLASS=A,MSGCLASS=H' \
+		'//STEP1    EXEC PGM=IEFBR14' \
+		'//')
+
+	local resp
+	resp=$(do_curl PUT \
+		-H "Content-Type: text/plain" \
+		-H "X-IBM-Intrdr-Mode: TEXT" \
+		-H "X-IBM-Intrdr-Lrecl: 80" \
+		-H "X-IBM-Intrdr-Recfm: F" \
+		--data-binary "$jcl" \
+		"${BASE_URL}/zosmf/restjobs/jobs")
+	split_response "$resp"
+
+	assert_http_status "200" "$HTTP_STATUS" "submit deck with padded null statement"
+	assert_json_field "$BODY" '.jobname' "NULLST" "job card found despite padded null statement"
+
+	# Purge this job to avoid clutter
+	local jn ji
+	jn=$(echo "$BODY" | jq -r '.jobname')
+	ji=$(echo "$BODY" | jq -r '.jobid')
+	if [ "$jn" != "null" ] && [ "$ji" != "null" ]; then
+		wait_for_output "$jn" "$ji" || true
+		do_curl DELETE "${BASE_URL}/zosmf/restjobs/jobs/${jn}/${ji}" >/dev/null 2>&1 || true
+	fi
+}
+
 test_submit_without_notify_gets_retcode() {
 	echo ""
 	echo "--- Submit Job: card without NOTIFY still reports a retcode (issue #307) ---"
@@ -1577,6 +1618,7 @@ fi
 test_submit_inline_jcl
 test_submit_inline_jcl_with_intrdr_headers
 test_submit_notify_sysuid_trailing_param
+test_submit_padded_null_statement
 test_submit_without_notify_gets_retcode
 test_submit_literal_notify_not_duplicated
 test_submit_notify_inside_programmer_name

